File: CS/CustomFilterPopup/CustomFilterPopup.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.UI;
using DevExpress.Xpf.Editors;
using DevExpress.Xpf.PivotGrid;
using DevExpress.Xpf.PivotGrid.Internal;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;

namespace CustomFilterPopup
{

    class PivotGridCustomFilter : ComboBoxEdit {
        public PivotGridCustomFilter() {
            PopupOpening += new OpenPopupEventHandler(PivotGridRadioFilter_PopupOpening);
            PopupClosed += new ClosePopupEventHandler(PivotGridRadioFilter_PopupClosed);
            InvertCommand = new DelegateCommand(Invert);
            EventManager.RegisterClassHandler(typeof(ComboBoxEditItem), ComboBoxEditItem.RequestBringIntoViewEvent, new RequestBringIntoViewEventHandler(OnRequestBringIntoView));
        }

        protected override void EndInitInternal(bool callBase) {
            var fieldHeaderControl = LayoutTreeHelper.GetVisualParents(Parent).First(i => i is FieldHeaderContentControl);
            var filterEdit = LayoutTreeHelper.GetVisualChildren(fieldHeaderControl).OfType<FilterPopupEdit>().FirstOrDefault();
            if (filterEdit != null) {
                var border = LayoutTreeHelper.GetVisualParents(filterEdit).First(i => i is Border) as Border;
                border.Visibility = Visibility.Collapsed;
            }

            base.EndInitInternal(callBase);
        }

        bool blockViewUpdates;

        void OnRequestBringIntoView(object sender, RequestBringIntoViewEventArgs e) {
            e.Handled = blockViewUpdates;
        }

        public ICommand InvertCommand { get; set; }

        public PivotGridField PivotField {
            get { return (PivotGridField)GetValue(PivotFieldProperty); }
            set { SetValue(PivotFieldProperty, value); }
        }

        // Using a DependencyProperty as the backing store for Field.  This enables animation, styling, binding, etc...
        public static readonly DependencyProperty PivotFieldProperty =
            DependencyProperty.Register("PivotField", typeof(PivotGridField), typeof(PivotGridCustomFilter), new UIPropertyMetadata(null));

        public static readonly DependencyProperty FilterButtonVisibleProperty =
    DependencyProperty.Register("FilterButtonVisible", typeof(Visibility), typeof(PivotGridCustomFilter), new UIPropertyMetadata(null));



        public void Invert() {
            blockViewUpdates = true;
            var unselectedItems = ListBox.ItemsSource.Cast<object>().OfType<PivotValueInfo>().Except(ListBox.SelectedItems.Cast<object>()).ToList();
            this.UnselectAllItems();
            unselectedItems.ForEach(i => SelectedItems.Add(i));
            blockViewUpdates = false;
        }

        void PivotGridRadioFilter_PopupOpening(object sender, OpenPopupEventArgs e) {
            if (PivotField == null) {
                e.Cancel = true;
                e.Handled = true;
                return;
            }
            this.BeginDataUpdate();
            Items.Clear();
            var uniqueItems = PivotField.GetUniqueValues().Select(v => new PivotValueInfo() { Value = v, DisplayText = PivotField.GetDisplayText(v) }).ToArray();
            var editValue = PivotField.FilterValues.ValuesIncluded;
            Items.AddRange(uniqueItems);
            this.EditValue = editValue;
            this.EndDataUpdate();
        }

        void PivotGridRadioFilter_PopupClosed(object sender, ClosePopupEventArgs e) {
            PivotGridControl pivot = PivotField.Parent as PivotGridControl;
            if (e.CloseMode == DevExpress.Xpf.Editors.PopupCloseMode.Cancel || PivotField == null || SelectedItems == null || pivot == null)
                return;
            var values = SelectedItems.OfType<PivotValueInfo>().Select(v => v.Value).ToArray();
            if (PivotField.FilterValues.FilterType == FieldFilterType.Included)
                PivotField.FilterValues.ValuesIncluded = values;
            else
                PivotField.FilterValues.ValuesExcluded = PivotField.GetUniqueValues().Except(values).ToArray();
        }
    }

    public class PivotValueInfo {
        public object Value { get; set; }
        public string DisplayText { get; set; }
    }

}

File: VB/CustomFilterPopup/CustomFilterPopup.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.Xpf.Editors
Imports System.Windows
Imports DevExpress.Xpf.PivotGrid
Imports System.Windows.Input
Imports DevExpress.Mvvm
Imports DevExpress.Xpf.Core.Native
Imports System.Windows.Controls
Imports System.Threading
Imports System.Windows.Threading
Imports DevExpress.Mvvm.UI
Imports DevExpress.Xpf.PivotGrid.Internal

Namespace HowToBindToMDB

    Friend Class PivotGridCustomFilter
        Inherits ComboBoxEdit


        Public Sub New()
            AddHandler PopupOpening, AddressOf PivotGridRadioFilter_PopupOpening
            AddHandler PopupClosed, AddressOf PivotGridRadioFilter_PopupClosed
            InvertCommand = New DelegateCommand(AddressOf Invert)
            EventManager.RegisterClassHandler(GetType(ComboBoxEditItem), ComboBoxEditItem.RequestBringIntoViewEvent, New RequestBringIntoViewEventHandler(AddressOf OnRequestBringIntoView))
        End Sub

        Protected Overrides Sub EndInitInternal(callBase As Boolean)
            Dim fieldHeaderControl = LayoutTreeHelper.GetVisualParents(Parent).First(Function(i) TypeOf i Is FieldHeaderContentControl)
            Dim filterEdit = LayoutTreeHelper.GetVisualChildren(fieldHeaderControl).OfType(Of FilterPopupEdit)().FirstOrDefault()
            If filterEdit IsNot Nothing Then
                Dim border = CType(LayoutTreeHelper.GetVisualParents(filterEdit).First(Function(i) TypeOf i Is Border), Border)
                border.Visibility = System.Windows.Visibility.Collapsed
            End If
            MyBase.EndInitInternal(callBase)
        End Sub

        Private blockViewUpdates As Boolean

        Private Sub OnRequestBringIntoView(ByVal sender As Object, ByVal e As RequestBringIntoViewEventArgs)
            e.Handled = blockViewUpdates
        End Sub

        Public Property InvertCommand() As ICommand

        Public Property PivotField() As PivotGridField
            Get
                Return CType(GetValue(PivotFieldProperty), PivotGridField)
            End Get
            Set(ByVal value As PivotGridField)
                SetValue(PivotFieldProperty, value)
            End Set
        End Property

        ' Using a DependencyProperty as the backing store for Field.  This enables animation, styling, binding, etc...
        Public Shared ReadOnly PivotFieldProperty As DependencyProperty = DependencyProperty.Register("PivotField", GetType(PivotGridField), GetType(PivotGridCustomFilter), New UIPropertyMetadata(Nothing))

        Public Shared ReadOnly FilterButtonVisibleProperty As DependencyProperty = DependencyProperty.Register("FilterButtonVisible", GetType(Visibility), GetType(PivotGridCustomFilter), New UIPropertyMetadata(Nothing))



        Public Sub Invert()
            blockViewUpdates = True
            Dim selectedValues = ListBox.SelectedItems.Cast(Of Object)().ToArray()
            Dim unselectedItems = ListBox.ItemsSource.Cast(Of Object)().OfType(Of PivotValueInfo)().Cast(Of Object)().Except(selectedValues).ToList()
            Me.UnselectAllItems()
            unselectedItems.ForEach(Sub(i) SelectedItems.Add(i))
            blockViewUpdates = False
        End Sub

        Private Sub PivotGridRadioFilter_PopupOpening(ByVal sender As Object, ByVal e As OpenPopupEventArgs)
            If PivotField Is Nothing Then
                e.Cancel = True
                e.Handled = True
                Return
            End If
            Me.BeginDataUpdate()
            Items.Clear()
            Dim uniqueItems = PivotField.GetUniqueValues().Select(Function(v) New PivotValueInfo() With {.Value = v, .DisplayText = PivotField.GetDisplayText(v)}).ToArray()
            Dim editValue = PivotField.FilterValues.ValuesIncluded
            Items.AddRange(uniqueItems)
            Me.EditValue = editValue
            Me.EndDataUpdate()
        End Sub

        Private Sub PivotGridRadioFilter_PopupClosed(ByVal sender As Object, ByVal e As ClosePopupEventArgs)
            Dim pivot As PivotGridControl = TryCast(PivotField.Parent, PivotGridControl)
            If e.CloseMode = DevExpress.Xpf.Editors.PopupCloseMode.Cancel OrElse PivotField Is Nothing OrElse SelectedItems Is Nothing OrElse pivot Is Nothing Then
                Return
            End If
            Dim values = SelectedItems.OfType(Of PivotValueInfo)().Select(Function(v) v.Value).ToArray()
            If PivotField.FilterValues.FilterType = FieldFilterType.Included Then
                PivotField.FilterValues.ValuesIncluded = values
            Else
                PivotField.FilterValues.ValuesExcluded = PivotField.GetUniqueValues().Except(values).ToArray()
            End If
        End Sub
    End Class

    Public Class PivotValueInfo
        Public Property Value() As Object
        Public Property DisplayText() As String
    End Class

End Namespace
